Log a warning message when failed to find kind for resource in garbage collector controller

zhouhaibing089 opened the pull request:
at this time, I do not think thirdparty api group version resources should be taken care by garbage collector controllers, and this line of call will fail actually: https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/garbagecollector/garbagecollector.go#L565, and as a result, the garbagecollector controller failed to start.

zhouhaibing089 committed:
warning instead error when search kinds for resources

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/39457)
<!-- Reviewable:end -->

zhouhaibing089 commented:
cc @caesarxuchao @sttts

sttts commented on pkg/master/thirdparty/thirdparty.go:
```diff
@@ -284,6 +284,7 @@ func (m *ThirdPartyResourceServer) InstallThirdPartyResource(rsrc *extensions.Th
 	m.genericAPIServer.HandlerContainer.Add(genericapi.NewGroupWebService(api.Codecs, path, apiGroup))
 
 	m.addThirdPartyResourceStorage(path, plural.Resource, thirdparty.Storage[plural.Resource].(*thirdpartyresourcedataetcd.REST), apiGroup)
+	registered.AddThirdPartyAPIGroupVersions(schema.GroupVersion{Group: group, Version: rsrc.Versions[0].Name})
```
Will the upper check see this registration? This is master, above is controller manager.

zhouhaibing089 replied:
it won't, I add this because there are some calls in the apiserver which did the same check as above, for example: https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/extensions/thirdpartyresourcedata/codec.go#L373.

but yeah, I did two things here, and I am still doing various of validation.

sttts replied:
Is anything broken in the apiserver due to not registering the thirdpartyresources? We are either missing a test or the whole registration is not necessary. In the later case, I would rather remove the whole registration logic for tprs.

zhouhaibing089 replied:
@sttts along with your changes to add verbs for thirdpartyresources, the namespace controller will be aware of them, so let's reconsider the problem at https://github.com/kubernetes/kubernetes/issues/37278.

When a namespace is going to be deleted, the namespace controller will delete all the deletable resources under that namespace, and in case it is a TPR resources, the delete options will failed to be decoded by a TPR gvk, and @caesarxuchao made a change in this pull request: https://github.com/kubernetes/kubernetes/pull/37328, if you look at the code, it checks whether the gvk is a TPR group version by calling `registered.IsThirdPartyAPIGroupVersion`, and that call depends on whether we register it or not.

lavalamp replied:
The stuff in the registered package is intended to be static, write-once on application startup, so I'm a little surprised to see this method...

lavalamp replied:
And answering the question, this line could only possibly affect the copy of `registered` in the apiserver binary. It has no effect on the copy of `registered` that the garbage collector would use.

If you add this here, probably there is a corresponding place where it needs to be removed.

lavalamp replied:
hm, it's pretty weird that that function existed but was never called in the server. Perhaps it was intended for clients constructing their own RESTMapper.

lavalamp replied:
Last thought: having this line here without explanation and without a test makes it hard to refactor this code into something understandable. @zhouhaibing089 it would be great if you could add a test :)

zhouhaibing089 replied:
@lavalamp thanks for your thoughts, i will do both: add an explanation and also a test.

sttts commented:
I don't think yet another special case for thirdparty resource handling is the wrong direction. We need watchable discovery on the apiserver. Then controllers can watch resources come and go, in order to update the restmappers.

In the namespace controller, dynamic discovery (= polling of the discovery endpoint) was implemented as a quick'n'dirty workaround.

zhouhaibing089 commented:
@sttts I agree with you that we need a dynamic rest mapper, and I do see there are some `TODO`s.

The reason I am thinking that we might want to ignore thirdparty resources is: the owner references things are not applied to the thirdpartyresources at this moment, given that there are still enhancements(issues like `kindsFor` does not work with TPRs) will come, I am little believing it should be fine to ignore them, and it actually resolves the problem which fails the garbage collector controller..

sttts commented:
The controller manager is a normal client, i.e. is has no knowledge and should not have knowledge about what thirdparty resources are. We cannot detect them. The `IsThirdPartyAPIGroupVersion` func is misleading (and actually shouldn't be there in the first place).

What is actual reason that we have deletable resource in that loop, but the restmapper doesn't know them?

zhouhaibing089 commented:
@sttts thanks for point it out, I've updated the code a little bit.

> What is actual reason that we have deletable resource in that loop, but the restmapper doesn't know them?

the restmapper could not tell the kind for thirdpartyresources, which means a `NoKindMatch` error will be returned, if the error is propagated up, the caller will `Fatalf` with the error message. I've updated the code to just log a warning message.

sttts commented:
But why doesn't it know about the Kind in the first place?

zhouhaibing089 commented:
@sttts I did not read deep into the code, I find the result by adding some logs..

at the first view, `registered.RESTMapper` builds a union mapper based on all the enabled versions, do you have any ideas on how these group versions are initialized from the perspective of client side?

sttts commented:
I haven't checked in the code. But it might be that we still have a static restmapper there, derived from the registered resources via pkg/apimachinery. The kubectl uses a dynamic restmapper nowadays which is fed through the discovery endpoints of the apiserver.

If you want to dive into that topic a bit, try to find out where the restmapper is coming from. I believe to have seen a comment somewhere – maybe in the controller manager – in the direction that discovery should be used, but isn't.

zhouhaibing089 commented:
@sttts thanks, I will go deeper at this front, I might figure out the TODOs at several followed pull requests, so far, are you good with the changes here?

sttts commented on pkg/controller/garbagecollector/garbagecollector.go:
```diff
@@ -564,7 +564,9 @@ func NewGarbageCollector(metaOnlyClientPool dynamic.ClientPool, clientPool dynam
 		if err != nil {
-			return nil, err
+			// NoKindMatch error would be returned for TPR resources
+			glog.Warningf("failed to get kind for %v", resource)
```
can we check for this specific error here?

zhouhaibing089 replied:
you mean adding the error info in the warning message?

zhouhaibing089 commented:
@sttts I just added the error type check, and I am sorry, the error is actually `NoResourceMatchError`, just verified on a running cluster.

zhouhaibing089 commented:
ping @lavalamp could you please take a look at the changes here? in short, the registered is not aware of TPRs, so I add the call to `AddThirdPartyAPIGroupVersions` when installing TPRs, is it the correct way or even necessary to do so?

sttts commented on pkg/controller/garbagecollector/garbagecollector.go:
```diff
@@ -564,7 +564,13 @@ func NewGarbageCollector(metaOnlyClientPool dynamic.ClientPool, clientPool dynam
+				// NoResourceMatchError would be returned for TPR
+				glog.Warningf("ignore NoResourceMatchError for %v", resource)
+				continue
+			} else {
```
no need for the else-clause

zhouhaibing089 replied:
@sttts don't we need to return the error when it is not of type NoResourceMatchError?

sttts replied:
sure. I mean only the else block around. The return should stay.

zhouhaibing089 replied:
got it and removed.

sttts commented on pkg/controller/garbagecollector/garbagecollector.go:
```diff
@@ -564,6 +564,11 @@ func NewGarbageCollector(metaOnlyClientPool dynamic.ClientPool, clientPool dynam
 		kind, err := gc.restMapper.KindFor(resource)
 		if err != nil {
+			if _, ok := err.(*meta.NoResourceMatchError); ok {
+				// NoResourceMatchError would be returned for TPR
```
A last thing: can you be a bit more verbose here, like:

```golang
// ignore NoResourceMatchErrors for nw because TPRs won't be registered and hence
// the RestMapper does not know about them. The deletableResources though are using discovery which
// included TPRs.
// TODO: use dynamic discovery for RestMapper and deletableResources
```

zhouhaibing089 replied:
thanks, done as you suggested. 👍

sttts commented:
@k8s-bot ok to test

sttts commented:
@zhouhaibing089 how relevant is this for 1.5?

zhouhaibing089 commented:
@sttts so two changes.


### Add third party api group version into registered

This is a known issue for both 1.5 and 1.4, in my opinion, if we are gonna to fix 1.5, we should pick this along with https://github.com/kubernetes/kubernetes/pull/37328.

### Changes about garbage collector controller

The problem happened on 1.5, too, so better to be picked.
